MoreLinq/UnFold.cs

```diff
@@ -0,0 +1,87 @@
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Returns a sequence generated by applying a state to the generator function, 
+        /// and from its result, determines if the sequence should have a next element and its value,
+        /// and the next state in the recursive call.
+        /// </summary>
+        /// <typeparam name="TState">Type of state elements.</typeparam>
+        /// <typeparam name="T">Type of the elements generated by the generator function.</typeparam>
+        /// <typeparam name="TResult">The type of the elements of the result sequence.</typeparam>
+        /// <param name="state">The initial state.</param>
+        /// <param name="generator">
+        /// Function that takes a state and produces an object containing all 
+        /// necessary data for the other functions (predicate, stateSelector, resultSelector).
+        /// </param>
+        /// <param name="predicate">
+        /// Function to determine if the unfolding should continue based on the current input.
+        /// </param>
+        /// <param name="stateSelector">
+        /// Function to select the state from the current input.
+        /// </param>
+        /// <param name="resultSelector">
+        /// Function to select the result from the current input.
+        /// </param>
+        /// <returns>A sequence containing the results generated by the resultSelector function.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<TResult> Unfold<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            if (generator == null) throw new ArgumentNullException(nameof(generator));
+            if (predicate == null) throw new ArgumentNullException(nameof(predicate));
+            if (stateSelector == null) throw new ArgumentNullException(nameof(stateSelector));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return UnfoldImpl(state, generator, predicate, stateSelector, resultSelector);
+        }
+
+
+        private static IEnumerable<TResult> UnfoldImpl<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            while (true)
+            {
+                var step = generator(state);
+
+                if (!predicate(step))
+                    yield break;
+
+                yield return resultSelector(step);
+                state = stateSelector(step);
```

Comment:
> How do you get back the result to yield for the current iteration from the state for the next that is step?

Well, you start with a seed *state*, not seed *value*:

```csharp
var squaresLowerThan100 =
    Unfold(new { x = 1, sq = 0 }, s => new { x = s.x+1, sq = s.x*s.x }, s => s.sq < 100, s => s.sq);
```

I think what is confusing me a lot is that I find that what is currently described as `T step` I would define as `TState state`. After all, it is `step` that carries all the information necessary to generate results, determine if iteration should continue, and generate the next iteration (if we assume `stateSelector` is usually a property access).

>>where one function is an identity
>
> Which function?

I originally wrote "[either resultSelector or generator are identity functions](https://github.com/morelinq/MoreLINQ/pull/260#issuecomment-284725887)". However, having a `generator` being an identity is not likely to be very useful (and certainly not common). This got me thinking I did not understand correctly what the parameters really mean, and I could not resolve my confusion.

BTW, what I really meant is `stateSelector` and `resultSelector` as identities.